### PR TITLE
Issue148 nautilus credentials update pt2

### PR DESCRIPTION
**Pull request by njackson2012**
I created a new branch, merged it with the old nautilus credentials branch, removed some offending changes, and would now like to put my stuff into master.

### CODE REVIEW

File: brp_admin/tests/test_brp_admin_protocols.py [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+    def test_that_all_of_users_nautilus_credentials_are_updated(self):
+        self.assertEqual(self.protocolusercredentials.data_source_password, "newPassword")
+        self.assertEqual(self.protocolusercredentials2.data_source_password, "newPassword")
+
+    def test_that_user_credentials_with_blank_passwords_are_left_alone(self):
+        self.assertNotEqual(self.protocolusercredentials5.data_source_password, "newPassword")
+
+    @pytest.mark.skip(reason="This particular test is impossible because not adding a user breaks the database.")
+    def test_that_blank_user_fields_do_not_throw_false_positives(self):
+        self.assertNotEqual(self.protocolusercredentials6.data_source_password, "newPassword")
```
Comment (seg1129) [inline brp_admin/tests/test_brp_admin_protocols.py]: This issue could be resolved by searching by 'data_source_username' field instead of user field - since that is actually what we are looking for and this field is not required. [Here](https://github.com/chop-dbhi/biorepo-portal/blob/5d36dac79531c801a680ee4bc979f027699ddd8b/api/models/protocols.py#L544) is the data model.

File: brp_admin/tests/test_brp_admin_protocols.py [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+            user=self.user2,
+            protocol_user=self.protocoluser,
+            data_source_username="unitName"
+            )
+        """
+
+            This case breaks the database...
+
+        ProtocolUserCredentials.objects.create(
+            protocol=self.protocol,
```
Comment (seg1129) [inline brp_admin/tests/test_brp_admin_protocols.py]: This is breaking because you are not setting the user field which has a not null constraint. We want to test to see if the username field is null. [Here](https://github.com/chop-dbhi/biorepo-portal/blob/5d36dac79531c801a680ee4bc979f027699ddd8b/api/models/protocols.py#L544) is the data model.

**Comment by seg1129** [comment]
There were two additional requirements added to this feature - please update issue #148 with additional requirements.

**Comment by seg1129** [comment]
If you want to clean up the setup for your test cases. [Here](https://github.com/chop-dbhi/ehb-service/blob/878621f6bd3d2d5fab5599a73175b0e30c01eb0b/ehb_service/apps/api/tests.py#L14) is an example of building test objects for a Django data model using fixtures.

**Comment by njackson2012** [comment]
I wanted to confirm the two changes to scope before documenting them.

1. Check to see if the username on a protocol user credential is mismatched from the user's username and do not change it if it is.
2. Alert the person making the change of any mismatched usernames so that they can resolve the conflict more easily.

**Comment by seg1129** [comment]
Close, here are the two additional requirements:
1. Check to see if the password on a protocol user credential and do not change it if it is.
2. Check to see if the username on a protocol user credential is mismatched from the user's username. If so alert the person making the change of any mismatched usernames so that they can resolve the conflict easily.

**Comment by njackson2012** [comment]
As per the latest code review, I am going to make the following changes:
**In protocol.py:**
- [ ] Explain in feature header that the password will be changed across every Nautilus user credential.
- [ ] Errors which happen through exceptions should get logged and the user will be informed of the logging.
- [ ] Order both query sets to match each other.
- [ ] Change the name of compSet to something better suited.
- [ ] Add comment describing the second use of the details list.
- [ ] Reorder the creation of packed message context data to put all appends at the end. This will eliminate the use of ambiguous index numbers.
- [ ] If no changes are made inform the user that no changes were made.
- [ ] Add collapsable list of unaltered user credentials.
- [ ] Add information format details to headers.

**In test_brp_admin_protocols_functional.py**
- [ ] Update titles of each test section to reflect the function being tested.
- [ ] Improve comments for databaseCases.

File: brp_admin/views/protocol.py [hunk 3]
```diff
@@ -29,6 +164,11 @@ def processProtocolUserForm(self, request):
         if protocolUserForm.is_valid():
             protocolUserForm.save()
         # if form is not valid - send errors to UI
+        """
```
Comment (seg1129) [inline brp_admin/views/protocol.py]: This comment is unrelated to issue #148. Please add details to issue #157, and reference this block of code in issue.

**Comment by njackson2012** [comment]
The following are cases we need to check for in unit testing for this feature:

**Data validation**
- [ ] The selected user must have their password updated for all qualifying Nautilus credentials.
- [ ] The selected user should have all of their non-nautilus credentials remain untouched.
- [ ] The selected user should have all qualifying Nautilus credentials with empty password fields remain untouched.
- [ ] The selected user should have all qualifying Nautilus credentials with datasource_usernames that do not match their CHOP assigned username remain untouched.
- [ ] Other users with Nautilus credentials should remain untouched.

**Error reporting**
- [ ] If the user selects a user but does not enter a password, they should receive an error stating that the password field is empty.
- [ ] If the user enters a password but does not select a user, they should receive an error stating that a user needs to be selected.
- [ ] If the user does not select a user and leaves the passwords field blank, they should get a concatenation of both errors.
